pilot/pkg/config/monitor/copilot_snapshot.go

```diff
@@ -0,0 +1,155 @@
+type CopilotClient interface {
+	copilotapi.IstioCopilotClient
+}
+
+// CopilotSnapshot provides a snapshot of configuration from the Cloud Foundry Copilot component.
+type CopilotSnapshot struct {
+	store             model.ConfigStore
+	client            CopilotClient
+	hostnameBlacklist []string
+	timeout           time.Duration
+	virtualServices   map[string]*model.Config
+}
+
+// NewCopilotSnapshot returns a CopilotSnapshot used for integration with Cloud Foundry.
+// The store is required to discover any existing gateways: the generated config will reference those gateways
+// The client is used to discover Cloud Foundry Routes from Copilot
+// The snapshot will not return routes for any hostnames which end with strings on the hostnameBlacklist
+func NewCopilotSnapshot(store model.ConfigStore, client CopilotClient, hostnameBlacklist []string, timeout time.Duration) *CopilotSnapshot {
+	return &CopilotSnapshot{
+		store:             store,
+		client:            client,
+		hostnameBlacklist: hostnameBlacklist,
+		timeout:           timeout,
+		virtualServices:   make(map[string]*model.Config),
+	}
+}
+
+// ReadConfigFiles returns a complete set of VirtualServices for all Cloud Foundry routes known to Copilot.
+// It may be used for the getSnapshotFunc when constructing a NewMonitor
+func (c *CopilotSnapshot) ReadConfigFiles() ([]*model.Config, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+
+	resp, err := c.client.Routes(ctx, new(copilotapi.RoutesRequest))
+	if err != nil {
+		return nil, err
+	}
+
+	gateways, err := c.store.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+
+	var gatewayNames []string
+	for _, gwConfig := range gateways {
+		gatewayNames = append(gatewayNames, gwConfig.Name)
+	}
+
+	filteredCopilotHostnames := c.removeBlacklistedHostnames(resp)
+
+	for _, hostname := range filteredCopilotHostnames {
+		if _, ok := c.virtualServices[hostname]; ok {
+			continue
+		}
+
+		c.virtualServices[hostname] = &model.Config{
+			ConfigMeta: model.ConfigMeta{
+				Type:    model.VirtualService.Type,
+				Version: model.VirtualService.Version,
+				Name:    fmt.Sprintf("route-for-%s", hostname),
+			},
+			Spec: &networking.VirtualService{
+				Gateways: gatewayNames,
+				Hosts:    []string{hostname},
+				Http: []*networking.HTTPRoute{
+					{
+						Route: []*networking.DestinationWeight{
+							{
+								Destination: &networking.Destination{
+									Name: hostname,
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	var cachedHostnames sort.StringSlice
+	for hostname := range c.virtualServices {
+		cachedHostnames = append(cachedHostnames, hostname)
+	}
+
+	cachedHostnames.Sort()
+
+	return c.collectVirtualServices(cachedHostnames, resp), nil
+}
+
+func (c *CopilotSnapshot) removeBlacklistedHostnames(resp *copilotapi.RoutesResponse) []string {
+	var filteredHostnames []string
```

Comment:
Initialize at the correct size please (`len(resp.Backends)` to avoid having to resize in worst case).